deepchem/utils/fake_data_generator.py

```diff
@@ -0,0 +1,107 @@
+  """
+  Generates a Fake Graph Dataset which can be used for testing or other purposes.
+
+  Note
+  ----
+  This fake data generator is based on torch_geometric.dataset.FakeDataset
+  class. The reason for having DeepChem's own implementation are two fold:
+  1) The current stable version (2.0.3) of torch_geometric's FakeDataset
+  doesn't support global features and kwargs.
+  2) The generated dataset in FakeDataset have same node feature distribution
+  irrespective of graph or node label which make it unsuitable to use during testing.
+  """
+
+  # TODO @arunppsg Once torch_geometric's version 2.0.4 is release which
+  # fixes above two issues, this can be depreciated.
+  def __init__(self,
+               num_graphs: int = 100,
+               avg_num_nodes: int = 10,
+               num_node_features: int = 5,
+               avg_degree: int = 4,
+               num_edge_features: int = 3,
+               num_classes: int = 2,
+               task: str = 'graph',
+               **kwargs):
+    """
+    Parameters
+    ----------
+    num_graphs: int, default 100
+      Number of graphs to generate
+    avg_num_nodes: int, default 10
+      Average number of nodes in a graph
+    num_node_features: int, default 5
+      Average number of node features in a graph
+    avg_degree: int, default 4
+      Average degree of the graph
+    num_edge_features: int, default 3
+      Average number of features in the edge
+    task: str, default 'graph'
+      Indicates node-level labels or graph-level labels
+    kwargs: optional
+      Additional graph attributes and their shapes , e.g. `global_features = 5`
+    """
+    self.num_graphs = num_graphs
+    self.avg_num_nodes = max(avg_num_nodes, avg_degree)
+    self.avg_degree = max(avg_degree, 1)
+    self.num_node_features = num_node_features
+    self.num_edge_features = num_edge_features
+    self.num_classes = num_classes
+    self.task = task
+    self.kwargs = kwargs
+
+  def generate_graphs(self) -> NumpyDataset:
+    graphs, labels = [], []
+    for i in range(self.num_graphs):
+      if self.task == 'graph':
+        label = random.randint(0, self.num_classes - 1)
+      elif self.task == 'node':
+        # TODO
+        pass
+      num_nodes = get_num_nodes(self.avg_num_nodes, self.avg_degree)
+      edge_index = get_edge_index(num_nodes, self.avg_degree)
+      num_edges = edge_index.shape[1]
+      node_features = np.random.rand(num_nodes, self.num_node_features) + label
+      edge_features = np.random.rand(num_edges, self.num_edge_features) + label
+      kwargs = {}
+      for feature_name, feature_shape in self.kwargs.items():
+        kwargs[feature_name] = np.random.rand(feature_shape) + label
+      graph = GraphData(node_features, edge_index, edge_features, **kwargs)
+      graphs.append(graph)
+      labels.append(label)
+
+    graph_dataset = NumpyDataset(X=graphs, y=labels)
+    return graph_dataset
+
+
+def get_num_nodes(avg_num_nodes: int, avg_degree: int) -> int:
```

Comment:
These should either be private functions or have docstrings

Reply:
Added docstrings since these functions can be used by other modules too.